### PR TITLE
Compact DNS logging - v5

### PR DESCRIPTION
**Pull request by jasonish**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
- https://redmine.openinfosecfoundation.org/issues/2199
- https://redmine.openinfosecfoundation.org/issues/1198
- https://redmine.openinfosecfoundation.org/issues/2086

Previous PR:
- https://github.com/OISF/suricata/pull/3125

Changes from last PR:
- Add authorities to the response record under the authorities key to prevent standalone records just for authorities.
- Include the query rrname in the response. Makes the response more useful on its own, and adds more contaxt to NXDOMAIN return types that don't have any answers.
- When logging responses, the response will only be logged if the query type is enabled for logging.
- Update Rust DNS to support v2 records (and remove the old format from Rust).

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- PR jasonish-pcap: https://buildbot.openinfosecfoundation.org/builders/jasonish-pcap/builds/269
- PR jasonish: https://buildbot.openinfosecfoundation.org/builders/jasonish/builds/622

cc: @glongo

### CODE REVIEW

File: src/output-json-alert.c [hunk 1]
```diff
@@ -184,6 +186,35 @@ static void AlertJsonDnp3(const Flow *f, json_t *js)
     return;
 }
 
+static void AlertJsonDns(const Flow *f, json_t *js)
+{
+#ifndef HAVE_RUST
```
Comment (victorjulien) [inline src/output-json-alert.c]: would be nice to have this for rust too. Not a requirement for this PR, but maybe something to try and get into 4.1?

File: doc/userguide/output/eve/eve-json-format.rst [hunk 2]
```diff
@@ -177,13 +177,29 @@ Event with extended logging:
...
+rather than an event for each answer.
+
+It's possible to customize how a dns answer will be logged with the following
+formats:
+
+* "detailed": "rrname", "rrtype", "rdata" and "ttl" fields are logged for each answer
+* "grouped": answers logged are aggregated by their type (A, AAAA, NS, ...)
+
+It will be still possible to use the old DNS logging format, you can control it
+with "version" option in dns configuration section.
```
Comment (norg) [inline doc/userguide/output/eve/eve-json-format.rst]: I would add 2 "the" :) with *the* version option in *the* dns config

File: doc/userguide/output/eve/eve-json-format.rst [hunk 3]
```diff
@@ -212,7 +228,68 @@ Example of a DNS query for the IPv4 address of "twitter.com" (resource record ty
...
+          "192.0.78.25"
+        ],
+        "CNAME": [
+          "suricata-ids.org"
+        ]
+      }
+  }
+
+
+Example of a old DNS answer with an IPv4 (resource record type 'A') return:
```
Comment (norg) [inline doc/userguide/output/eve/eve-json-format.rst]: a*n* old

File: suricata.yaml.in [hunk 4]
```diff
@@ -203,13 +203,35 @@ outputs:
...
+            #enabled: no
+
+            # Control logging of requests and responses:
+            # - requests: enable logging of DNS queries
+            # - responses: enable logging of DNS answers
+            # By default both requests and responses are logged.
+            #requests: no
+            #responses: no
+
+            # Format of answer logging:
```
Comment (norg) [inline suricata.yaml.in]: same as above

File: doc/userguide/output/eve/eve-json-format.rst (hunk 2 above)
Comment (jasonish) [reply doc/userguide/output/eve/eve-json-format.rst]: Fixed.

File: doc/userguide/partials/eve-log.yaml [hunk 5]
```diff
@@ -86,6 +86,37 @@ outputs:
...
+            #enabled: no
+
+            # Control logging of requests and responses:
+            # - requests: enable logging of DNS queries
+            # - responses: enable logging of DNS answers
+            # By default both requests and responses are logged.
+            #requests: no
+            #responses: no
+
+            # Format of answer logging:
```
[the remark this answers is not in the document]
Comment (jasonish) [reply doc/userguide/partials/eve-log.yaml]: Change to "Answer logging format:"

**Comment by jasonish** [comment]
@victorjulien I've added a new skip field to suricata-verify that will skip some dns tests if eve.dns version 2 is being used. Should allow this branch to pass suricata-verify now without needing to use the dns-v2 branch of suricata-verify.

**Comment by victorjulien** [comment]
This branch rebased on suricata-verify master:
<pre>
$ ../suricata-verify/run.py
===> alert-testmyids: OK
===> alert-testmyids-not-established: OK
===> dnp3: OK
===> dnp3-dnp3_data-alert: OK
===> dnp3-dnp3_func-alert: OK
===> dns-eve-v2-udp-dig-a-www-suricata-ids-org: OK
===> dns-json-log: FAIL: got exit code 1, expected 0
===> dns-lua-rules: SKIPPED: Known issue with feature RUST
===> dns-tcp-multirequest-buffer-1: FAIL: got exit code 1, expected 0
===> dns-tcp-ts-gap: SKIPPED: not for outputs.1.eve-log.types.2.dns.version = 2
===> dns-tcp-www-google-com: FAIL: got exit code 1, expected 0
===> dns-udp-dig-a-www-suricata-ids-org: SKIPPED: not for outputs.1.eve-log.types.2.dns.version = 2
===> dns-udp-dns-log-unanswered: SKIPPED: not for feature RUST
===> dns-udp-double-request-response: FAIL: got exit code 1, expected 0
===> dns-udp-eve-log-aaaa-only: FAIL: got exit code 1, expected 0
===> dns-udp-eve-log-answer-only: FAIL: got exit code 1, expected 0
===> dns-udp-eve-log-mx-only: FAIL: got exit code 1, expected 0
===> dns-udp-eve-log-query-only: FAIL: got exit code 1, expected 0
===> dns-udp-eve-log-txt: SKIPPED: not for outputs.1.eve-log.types.2.dns.version = 2
===> dns-udp-nxdomain-soa: FAIL: got exit code 1, expected 0
===> dns-udp-unsolicited-response: SKIPPED: not for feature RUST
===> dns-udp-z-flag-fp: OK
===> eve-alert-metadata-defaults: FAIL: got exit code 1, expected 0
===> eve-alert-metadata-enable-rule: FAIL: got exit code 1, expected 0
===> eve-alert-metadata-off: FAIL: got exit code 1, expected 0
===> eve-dns: FAIL: got exit code 1, expected 0
===> eve-metadata: FAIL: got exit code 1, expected 0
===> filestore-v2.1-forced: OK
===> filestore-v2.2-forced-with-open-files: OK
===> filestore-v2.3-fserror: SKIPPED: Test fails when run as root
===> filestore-v2.4-forced-with-meta: OK
===> filestore-v2.5-both-enabled: OK
===> http-xff-eve-forward-extra-data: OK
===> http-xff-eve-forward-overwrite: OK
===> http-xff-eve-reverse-extra-data: OK
===> http-xff-eve-reverse-overwrite: OK
===> http-xff-unified2: SKIPPED: requires script returned false
===> linktype-228: OK
===> lua-output-dns: OK
===> lua-output-http: OK
===> lua-output-smtp: OK
===> output-eve-fileinfo: OK
===> output-pcap-log: OK
===> output-tcp-data: OK
===> proto-mismatch-http-ssh: OK
===> show-help: OK
===> single-dns-request: FAIL: got exit code 1, expected 0
===> smtp: OK
===> test-config-empty-rule-file: FAIL: got exit code 1, expected 0
===> tls: OK
===> tls-fingerprint-alert: OK
===> tls-json-output-ids: OK
===> tls-json-output-ips: OK

PASSED:  29
FAILED:  16
SKIPPED: 8
ERROR: Job failed: exit code 1
</pre>

This is supposed to pass, right?

**Comment by victorjulien** [comment]
Replaced by #3287

**Pull request closed without merging**